p2p/src/sync/mod.rs

```diff
@@ -89,445 +97,485 @@ where
+                })
             }
-            Err(err) => Err(err),
-        };
+            // Due to the fact that p2p is split into several tasks, it is possible to send a
+            // request/response after a peer is disconnected, but before receiving the disconnect
+            // event. Therefore this error can be safely ignored.
+            P2pError::PeerError(PeerError::PeerDoesntExist) => Ok(()),
+            // Some of these errors aren't technically fatal, but they shouldn't occur in the sync
+            // manager.
+            e @ (P2pError::DialError(_)
+            | P2pError::ConversionError(_)
+            | P2pError::PeerError(_)
+            | P2pError::NoiseHandshakeError(_)
+            | P2pError::PublishError(_)
+            | P2pError::InvalidConfigurationValue(_)
+            | P2pError::ChainstateError(_)) => Err(e),
+            // Fatal errors, simply propagate them to stop the sync manager.
+            e @ (P2pError::ChannelClosed
+            | P2pError::SubsystemFailure
+            | P2pError::StorageFailure(_)) => Err(e),
+        }
+    }
 
-        let score = match result {
-            Ok(_) => 0,
-            Err(e) => match e {
-                ChainstateError::FailedToInitializeChainstate(_) => 0,
-                ChainstateError::ProcessBlockError(err) => err.ban_score(),
-                ChainstateError::FailedToReadProperty(_) => 0,
-                ChainstateError::BootstrapError(_) => 0,
-            },
-        };
+    /// Sends a header list request to the given peer.
+    async fn request_headers(&mut self, peer: T::PeerId) -> Result<()> {
+        let locator = self.chainstate_handle.call(|this| this.get_locator()).await??;
+        debug_assert!(locator.len() <= self.p2p_config.msg_max_locator_count.clone().into());
+
+        self.messaging_handle
+            .send_message(
+                peer,
+                SyncMessage::HeaderListRequest(HeaderListRequest::new(locator)),
+            )
+            .map(|_| ())
+    }
 
-        if score > 0 {
-            // TODO: better abstraction over channels
-            let (tx, rx) = oneshot_nofail::channel();
-            self.tx_peer_manager
-                .send(PeerManagerEvent::AdjustPeerScore(peer_id, score, tx))
-                .map_err(P2pError::from)?;
-            let _ = rx.await.map_err(P2pError::from)?;
+    /// Sends a block list request to the given peer.
+    ///
+    /// The number of headers sent equals to `P2pConfig::requested_blocks_limit`, the remaining
+    /// headers are stored in the peer context.
+    fn request_blocks(&mut self, peer: T::PeerId, mut headers: Vec<BlockHeader>) -> Result<()> {
+        let peer_state = self
+            .peers
+            .get_mut(&peer)
+            .ok_or(P2pError::PeerError(PeerError::PeerDoesntExist))?;
+
+        debug_assert!(peer_state.known_headers.is_empty());
+        if headers.len() > self.p2p_config.max_request_blocks_count.clone().into() {
```

Comment:
OK... this is becoming a pattern. In many places you do clone then into unnecessarily. Please look into them all and simply use deref